File: go/1976.go
package main

func countPaths(n int, roads [][]int) int {
	g := make([][][2]int, n)
	for _, road := range roads {
		x, y, d := road[0], road[1], road[2]
		g[x] = append(g[x], [2]int{y, d})
		g[y] = append(g[y], [2]int{x, d})
	}
	counts := make([]int, n)
	dists := make([]int, n)
	dones := make([]bool, n)
	for i := range dists {
		dists[i] = -1
	}
	dists[0] = 0
	counts[0] = 1
	dones[0] = true
	minNode := 0
	for i := 0; i < n-1; i++ {
		for _, nxtNode := range g[minNode] {
			nxt, d := nxtNode[0], nxtNode[1]
			if dones[nxt] {
				continue
			}
			if dists[nxt] == -1 || d+dists[minNode] < dists[nxt] {
				counts[nxt] = counts[minNode]
				dists[nxt] = d + dists[minNode]
			} else if d+dists[minNode] == dists[nxt] {
				counts[nxt] = (counts[nxt] + counts[minNode]) % 1_000_000_007
			}
		}
		minNode = -1
		for j := 0; j < n; j++ {
			if !dones[j] && dists[j] != -1 && (minNode == -1 || dists[j] < dists[minNode]) {
				minNode = j
			}
		}
		dones[minNode] = true
		// fmt.Printf("minNode: %v\n", minNode)
		// fmt.Printf("dists: %v\n", dists)
		// fmt.Printf("counts: %v\n", counts)
	}

	return counts[n-1]
}

File: python/1976.py
from typing import List

class Solution:
    def countPaths(self, n: int, roads: List[List[int]]) -> int:
        g = [[] for _ in range(n)]
        for x, y, d in roads:
            g[x].append([y, d])
            g[y].append([x, d])
        dones = [False] * n
        counts = [0] * n
        dists = [-1] * n
        dones[0] = True
        counts[0] = 1
        dists[0] = 0
        min_node = 0
        for _ in range(n-1):
            for nxt, d in g[min_node]:
                if dones[nxt]: continue
                if dists[nxt] == -1 or d + dists[min_node] < dists[nxt]:
                    dists[nxt] = d + dists[min_node]
                    counts[nxt] = counts[min_node]
                elif d + dists[min_node] == dists[nxt]:
                    counts[nxt] = (counts[min_node] + counts[nxt]) % 1_000_000_007
            min_node = -1
            for i in range(n):
                if (not dones[i]) and dists[i] != -1 and (min_node == -1 or dists[i] < dists[min_node]):
                    min_node = i
            dones[min_node] = True
        return counts[-1]

            